File: neetcode-250/linked-list/206-reverse-linked-list.ts
import { ListNode } from "./listNode";

function reverseList(head: ListNode | null): ListNode | null {
	if (!head) {
		return null;
	}

	let next = head.next;
	head.next = null;

	while (next) {
		let temp = next.next;
		next.next = head;
		head = next;
		next = temp;
	}

	return head;
}

function reverseListRecursively(head: ListNode | null): ListNode | null {
	if (head?.next) {
		const temp = head;
		doReverse(head);
		temp.next = null;
	}
	return head;

	function doReverse(node: ListNode): ListNode {
		if (!node.next) {
			head = node;
			return node;
		}

		const next = doReverse(node.next);
		next.next = node;
		return node;
	}
}

File: neetcode-250/linked-list/206-reverse-linked-list.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return None

        def reverse(prev, node):
            next = node.next
            node.next = prev
            return reverse(node, next) if next else node

        return reverse(None, head)


class Solution2:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return head

        next = head.next
        head.next = None

        while next:
            temp = next.next
            next.next = head
            head = next
            next = temp

        return head
